File: src/main/web/problems.js
const util = require("../util");
const register = require("../util/register");
const Problem = require("../model").Problem;

register.post("/getProblems", "admin", async _ => {
    return await Problem.allJSON();
});

register.post("/getProblem", "admin", async (params) => {
    const id = params.id;
    var problem = await Problem.construct(id);
    return await problem.toJSONFull();
});

register.post("/deleteProblem", "admin", async (params) => {
    const id = params.id;
    await util.db.deleteKey(`/problems/${id}`);
    return "ok";
});

register.post("/editProblem", "admin", async (params) => {
    const id = params.id;
    let problem = id ? await Problem.construct(id): new Problem();
    
    problem.title       = params.title;
    problem.description = params.description;
    problem.statement   = params.statement;
    problem.input       = params.input;
    problem.output      = params.output;
    problem.constraints = params.constraints;
    problem.samples     = params.samples;
    problem.tests       = 0;

    await problem.save();

    return problem.id;
});

File: src/main/web/problems.py
from code.util import register
from code.util.db.problem import Problem, Datum
import json

def getProblems(params, setHeader, user):
    return Problem.allJSON()

def getProblem(params, setHeader, user):
    id = params["id"]
    return Problem.get(id).toJSONFull()

def deleteProblem(params, setHeader, user):
    id = params["id"]
    Problem.get(id).delete()
    return "ok"

def editProblem(params, setHeader, user):
    id = params.get("id")
    problem = Problem.get(id) or Problem()

    problem.title       = params["title"]
    problem.description = params["description"]
    problem.statement   = params["statement"]
    problem.input       = params["input"]
    problem.output      = params["output"]
    problem.constraints = params["constraints"]
    problem.samples     = int(params["samples"])

    testData            = json.loads(params["testData"])
    problem.testData    = [Datum(d["input"], d["output"]) for d in testData]
    problem.tests       = len(testData)

    problem.save()

    return problem.id

register.post("/getProblems", "admin", getProblems)
register.post("/getProblem", "admin", getProblem)
register.post("/deleteProblem", "admin", deleteProblem)
register.post("/editProblem", "admin", editProblem)
